base/gmp.jl

```diff
@@ -30,43 +30,85 @@ const GMP_BITS_PER_LIMB = gmp_bits_per_limb()
 # be used whenever mp_limb_t is in the signature of ccall'ed GMP functions.
 if GMP_BITS_PER_LIMB == 32
     typealias Limb UInt32
+    typealias SLimbMax Union{Int8, Int16, Int32}
+    typealias ULimbMax Union{UInt8, UInt16, UInt32}
 elseif GMP_BITS_PER_LIMB == 64
     typealias Limb UInt64
+    typealias SLimbMax Union{Int8, Int16, Int32, Int64}
+    typealias ULimbMax Union{UInt8, UInt16, UInt32, UInt64}
 else
     error("GMP: cannot determine the type mp_limb_t (__gmp_bits_per_limb == $GMP_BITS_PER_LIMB)")
 end
 
+immutable BigInt <: Integer
+    size::Int
+    d::Vector{Limb}
+end
+
+BigInt() = BigInt(0)
 
-type BigInt <: Integer
+type MPZ <: Integer
     alloc::Cint
     size::Cint
     d::Ptr{Limb}
-    function BigInt()
-        b = new(zero(Cint), zero(Cint), C_NULL)
-        ccall((:__gmpz_init,:libgmp), Void, (Ptr{BigInt},), &b)
-        finalizer(b, _gmp_clear_func)
-        return b
-    end
+
+    MPZ() = new(zero(Cint), zero(Cint), C_NULL)
+    # global objects can't be initialized via libgmp in a constructor, hence the init() function
+end
+
+function init(z::MPZ)
+    ccall((:__gmpz_init,:libgmp), Void, (Ptr{MPZ},), &z)
+    # finalizing is probably unnecessary if init'ed MPZ object are only const global
+    finalizer(z, cglobal((:__gmpz_clear, :libgmp)))
+end
+
+# to wrap an existing BigInt x as an MPZ, we call a global MPZ object _W on x: _W(x)
+# the syntax MPZ(x) would work, but would allocate a new MPZ on the heap each time
+function (z::MPZ)(x::BigInt)
+    z.alloc = z.size = x.size
+    z.d = pointer(x.d)
+    z
+end
+
+function BigInt(z::MPZ)
+    len = abs(z.size)
+    x = BigInt(z.size, Vector{Limb}(len))
+    unsafe_copy!(pointer(x.d), z.d, len)
+    x
 end
 
-_gmp_clear_func = C_NULL
+# _U, _V, _W are read-only shallow wrappers of BigInt objects ("const" arguments in libgmp)
+const _U = MPZ()
+const _V = MPZ()
+const _W = MPZ()
+
+# _X, _Y, _Z are read-write MPZ objects
+const _X = MPZ()
+const _Y = MPZ()
+const _Z = MPZ()
```

Comment:
Are there other similar cases in Base which have taken the lead on thread-safety, whose example I could follow?